### PR TITLE
fix: prevent concurrency issues in in-app listeners

### PR DESCRIPTION
**Pull request by mrehan27**
fixes: https://github.com/customerio/customerio-android/issues/245

### Changes

- Replaced for each loop with iterator to prevent concurrency issues

### CODE REVIEW

File: messaginginapp/src/main/java/io/customer/messaginginapp/gist/presentation/GistSdk.kt [hunk 1]
```diff
@@ -251,6 +257,14 @@ object GistSdk : Application.ActivityLifecycleCallbacks {
         return sharedPreferences.getString(SHARED_PREFERENCES_USER_TOKEN_KEY, null)
     }
 
+    private fun forEachListener(action: (MutableIterator<GistListener>) -> Unit) {
```
Comment (Shahroz16) [inline messaginginapp/src/main/java/io/customer/messaginginapp/gist/presentation/GistSdk.kt:260]: lets add a unit test for this method?

**Review by Shahroz16 — COMMENTED** [review]
We could probably move to another thread-collection to instead of `List` to avoid this but happy either way

File: messaginginapp/src/main/java/io/customer/messaginginapp/gist/presentation/GistSdk.kt (hunk 1 above)
Comment (mrehan27) [reply messaginginapp/src/main/java/io/customer/messaginginapp/gist/presentation/GistSdk.kt:260]: Yeah, I'm writing down some tests for this, will push once I'm able to fail it for the current code and pass for the new changes.

**Comment by mrehan27** [comment]
I tried different approaches and Shahroz's [suggestion for trying thread safe collections](https://github.com/customerio/customerio-android/pull/246#pullrequestreview-1586735619) looked the best option. I have used [CopyOnWriteArrayList](https://developer.android.com/reference/java/util/concurrent/CopyOnWriteArrayList) which works well and added tests. The tests passes with current implementation and fail if we run them on previous implementation.

**Review by Shahroz16 — APPROVED** [review]
Thank you for adding tests 👍

File: messaginginapp/src/sharedTest/java/io/customer/messaginginapp/GistSdkListenersTest.kt [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+import java.util.concurrent.TimeUnit
+import kotlin.concurrent.thread
+import org.amshove.kluent.internal.assertEquals
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
+internal class GistSdkListenersTest : BaseTest() {
+    @Test
+    fun processAndRemoveListenersIndividually_givenConcurrentModification_expectSuccessfulCompletion() {
```
Comment (levibostian) [inline messaginginapp/src/sharedTest/java/io/customer/messaginginapp/GistSdkListenersTest.kt:18]: Could you add a comment to this function explaining what edge case the test function is testing against (why this PR was created) and a small description about how this test function is designed? 

I don't expect anything long.

Comment (mrehan27) [reply messaginginapp/src/sharedTest/java/io/customer/messaginginapp/GistSdkListenersTest.kt:28]: Added [docs](https://github.com/customerio/customerio-android/pull/246/commits/19fce997bd0bd4953128031ebf457ecb515c6578)

File: messaginginapp/src/sharedTest/java/io/customer/messaginginapp/GistSdkListenersTest.kt [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+        assertEquals(
+            expected = 0L,
+            actual = threadsCompletionLatch.count,
+            message = "Threads did not complete within the timeout"
+        )
+    }
+
+    @Test
+    fun processAndClearListenersAllAtOnce_givenConcurrentModification_expectSuccessfulCompletion() {
```
[the remark this answers is not in the document]
Comment (mrehan27) [reply messaginginapp/src/sharedTest/java/io/customer/messaginginapp/GistSdkListenersTest.kt:83]: Yes, because: 

- `removeListener` and `clearListeners` are two different functions in `GistSdk`
- They are independent and do not reuse each other
- They have different logics (one removes listener instantly while the latter loops over all listeners)

File: messaginginapp/src/sharedTest/java/io/customer/messaginginapp/GistSdkListenersTest.kt (hunk 2 above)
Comment (levibostian) [reply messaginginapp/src/sharedTest/java/io/customer/messaginginapp/GistSdkListenersTest.kt:28]: Great docs. Thanks!

File: messaginginapp/src/sharedTest/java/io/customer/messaginginapp/GistSdkListenersTest.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (levibostian) [reply messaginginapp/src/sharedTest/java/io/customer/messaginginapp/GistSdkListenersTest.kt:83]: Sorry, I didn't realize there were 2 different functions. Yes, having 2 test functions is good then.
